### DISCUSSION TITLE
Building Updates from Source

### DISCUSSION THREAD

**Discussion by ryneeverett** [Divested-Mobile/DivestOS-Build#327] [Q&A] [answered]
Is there any information or resources on building and serving updates from source? The [build guide](https://divestos.org/pages/build) seems to make no mention of it.

Further, there seems to be no way to change the update servers in "Updater" settings. I take it one would need to patch the `DOS_OTA_SERVER_*` variables before building in order to change these?

**Comment by SkewedZeppelin** [comment, marked as answer]
The Updater paths need to be pointed to your own: https://github.com/Divested-Mobile/DivestOS-Build/blob/master/Scripts/init.sh#L86-L94

The server parts are here: https://github.com/Divested-Mobile/mirror.divestos.org

Just have all those files in the in a web root, edit mirrorutils/config.php, and then builds go into builds/LineageOS/device

See here for an example push: https://github.com/Divested-Mobile/DivestOS-Build/blob/master/Scripts/Common/Functions.sh#L487-L494

**Reply by ryneeverett** [reply]
Thanks, I think that pretty well covers how to serve updates, but how do you build `incrementals/`?

**Reply by SkewedZeppelin** [reply]
turn on https://github.com/Divested-Mobile/DivestOS-Build/blob/master/Scripts/init.sh#L68

I stopped generating them because even on my 7950X it can take longer than the actual compile to generate since the process is single threaded and RAM intensive.

ie. generating a given delta can take up like 20GB of RAM

GrapheneOS processes them as an after step on multiple machines in batches of 4 last I asked

**Reply by ryneeverett** [reply]
I had assumed incremental was the only way for OTA to work, but from what you're saying it sounds like it is possible to do OTA updates by serving the full zipped image and that is how DivestOS currently works. Do I have that right?

**Reply by SkewedZeppelin** [reply]
correct, incrementals are entirely optional
